### PR TITLE
Ensure handle to temporary is closed before temporary file expires.

### PR DESCRIPTION
**Pull request by ezyang**
Just the last commit.  This is a show-stopping bugfix for Windows which should go into 1.24.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
LGTM.

File: cabal-install/Distribution/Client/SrcDist.hs [hunk 1]
```diff
@@ -179,6 +181,8 @@ allPackageSourceFiles verbosity packageDir = do
...
 
   -- Run setup sdist --list-sources=TMPFILE
-  ret <- doListSources `catchIO` (\e -> onFailedListSources e >> return [])
-  removeExistingFile file
-  return ret
+  r <- doListSources `catchIO` (\e -> onFailedListSources e >> return [])
+  -- Ensure that we've closed the 'readFile' handle before we exit the
+  -- temporary directory.
+  _ <- evaluate (length r)
+  return r
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/SrcDist.hs]: Trivial point: `evaluate` does return the value, so don't need the separate return.

File: cabal-install/Distribution/Client/SrcDist.hs [hunk 2]
```diff
@@ -179,6 +181,8 @@ allPackageSourceFiles verbosity packageDir = do
           "Exception was: " ++ show e
 
   -- Run setup sdist --list-sources=TMPFILE
-  ret <- doListSources `catchIO` (\e -> onFailedListSources e >> return [])
-  removeExistingFile file
-  return ret
+  r <- doListSources `catchIO` (\e -> onFailedListSources e >> return [])
+  -- Ensure that we've closed the 'readFile' handle before we exit the
+  -- temporary directory.
+  _ <- evaluate (length r)
```
Comment (dcoutts) [inline cabal-install/Distribution/Client/SrcDist.hs]: I'd put the forcing at the source, which is:

    fmap lines . readFile $ file

**Comment by dcoutts** [comment]
LGTM

**Comment by ezyang** [comment]
Merged to master and 1.24.

**Pull request closed without merging**